components/scream/src/share/io/scorpio_input.cpp

```diff
@@ -0,0 +1,277 @@
+ * degrees of freedom assigned to each core and the correct io decomposition.
+ */
+  using namespace scream;
+  using namespace scream::scorpio;
+
+  // Cycle through all fields and register.
+  for (auto const& name : m_fields)
+  {
+    auto fmap = m_field_repo->get_field(name, m_grid_name);
+    auto& fid  = fmap.get_header().get_identifier();
+    // Determine the IO-decomp and construct a vector of dimension ids for this variable:
+    std::vector<std::string> vec_of_dims = get_vec_of_dims(fid.get_layout().tags());
+    std::string io_decomp_tag           = get_io_decomp(vec_of_dims);
+    get_variable(m_filename, name, name, vec_of_dims.size(), vec_of_dims, PIO_REAL, io_decomp_tag);  // TODO  Need to change dtype to allow for other variables.  Currently the field_repo only stores Real variables so it is not an issue, but in the future if non-Real variables are added we will want to accomodate that.
+    //TODO: Should be able to simply inquire fromt he netCDF the dimensions for each variable.
+  }
+  if (m_is_rhist) { get_variable(m_filename,"avg_count","avg_count",1,{"cnt"}, PIO_INT, "cnt"); }
+} // register_variables
+/* ---------------------------------------------------------- */
+std::vector<std::string> AtmosphereInput::get_vec_of_dims(const std::vector<FieldTag>& tags)
+{
+/* Given a set of dimensions in field tags, extract a vector of strings for those dimensions to be used with IO */
+  std::vector<std::string> vec_of_dims;
+  for (auto& tag_ii : tags)
+  {
+    vec_of_dims.push_back(e2str(tag_ii));
+  }
+  std::reverse(vec_of_dims.begin(),vec_of_dims.end()); // TODO: Reverse order of dimensions to match flip between C++ -> F90 -> PIO, may need to delete this line when switching to fully C++/C implementation.
+  return vec_of_dims;
+}
+/* ---------------------------------------------------------- */
+std::string AtmosphereInput::get_io_decomp(const std::vector<std::string>& vec_of_dims)
+{
+/* Given a vector of field dimensions, create a unique decomp string to register with I/O/
+ * Note: We are hard-coding for only REAL input here.  TODO: would be to allow for other dtypes
+ */
+  std::string io_decomp_tag = "Real";
+  for (auto& tag_ii : vec_of_dims)
+  {
+    io_decomp_tag += "-" + tag_ii;
+  }
+  return io_decomp_tag;
+}
+/* ---------------------------------------------------------- */
+void AtmosphereInput::register_views()
+{
+  using namespace scream;
+  using namespace scream::scorpio;
+
+  // Cycle through all fields and register.
+  for (auto const& name : m_fields)
+  {
+    auto fmap = m_field_repo->get_field(name, m_grid_name);
+    // If the "averaging type" is instant then just need a ptr to the view.
+    auto view_d = fmap.get_view();
+    // Create a local copy of view to be stored by input stream.
+    auto view_copy_h = Kokkos::create_mirror_view( view_d );
+    view_type view_copy("",view_copy_h.extent(0));
+    Kokkos::deep_copy(view_copy, view_d);
+    m_view_local.emplace(name,view_copy);
```

Comment:
Some users might need the host view from a field. Instead of creating two host mirrors, you should use the one inside fmap (btw, fmap is a field, not a map, the name is probably misleading). You could do this:

`m_view_local.emplace(name,fmap.get_view<Host>());`

Now, what's inside your m_view_local is the same that is in the host mirror inside fmap (saving a host view). And you can always sync them with `fmap.sync_to_host()` and `fmap.sync_to_dev()`.